crates/storage/src/connection/event.rs

```diff
@@ -343,13 +383,44 @@ impl KeyFilter for V03KeyFilter {
                     key_fts_expression.push(')');
                 }
             });
-            Some(KeyFilterResult {
-                base_query: " CROSS JOIN starknet_events_keys_03 ON starknet_events.rowid = starknet_events_keys_03.rowid",
-                where_statement: "starknet_events_keys_03.keys MATCH :events_match",
-                param: (":events_match", key_fts_expression),
-            })
-        } else {
-            None
+
+            Some(key_fts_expression)
+        };
+
+        Self { key_fts_expression }
+    }
+}
+
+impl KeyFilter for V03KeyFilter {
+    fn count(&self, tx: &Transaction<'_>) -> anyhow::Result<Option<usize>> {
+        match &self.key_fts_expression {
+            None => Ok(None),
+            Some(key_fts_expression) => {
+                let count: usize = tx.inner().query_row(
+                    "SELECT COUNT(1) FROM starknet_events_keys_03 WHERE keys MATCH :events_match",
+                    [&key_fts_expression],
+                    |row| row.get(0),
+                )?;
+                Ok(Some(count))
+            }
+        }
+    }
+
+    fn apply(&self, strategy: QueryStrategy) -> Option<KeyFilterResult<'_>> {
+        match self.key_fts_expression.as_ref() {
+            None => None,
+            Some(key_fts_expression) => {
+                let base_query = match strategy {
+                    QueryStrategy::BlockRangeFirst => " CROSS JOIN starknet_events_keys_03 ON starknet_events.rowid = starknet_events_keys_03.rowid",
+                    QueryStrategy::KeysFirst => " INNER JOIN starknet_events_keys_03 ON starknet_events.rowid = starknet_events_keys_03.rowid",
```

Comment:
Will the sqlite planner not sometimes still flip this order around if we accidentily upload a snapshot that has been analyzed?

Reply:
Yeah, that might be a problem... I don't think we'll upload analyzed snapshots though, but we could still make sure that it works well if you have an analyzed db... Will look into this.

Reply:
I don't know how to do that without reshuffling the code a lot -- the problem is that forcing the keys-first order we'd need to use CROSS JOIN with the keys table showing up first in the query...

Reply:
A problem for another day then.